File: src/stores/agentPrompts.ts
import { invoke } from "@tauri-apps/api/core";
import { create } from "zustand";
import { buildSendNowSchedule, schedulesToPruneForSend } from "../lib/agentPromptSend";
import { useAgentSchedulesStore } from "./agentSchedules";

/**
 * A prompt collected for a project without a tab binding. It lives in the
 * state dir's `agent_prompts.json`, keyed by scope, and becomes a schedule only
 * when aimed at an agent tab (`queuePromptForTab`, or the schedule dialog with
 * the text prefilled).
 */
export interface ProjectAgentPrompt {
  id: string;
  message: string;
  created_at: string;
  updated_at: string;
  /** Lowercase tokens the library is searched by (`lib/agentPromptTags`). */
  tags?: string[];
}

/**
 * A collected prompt after it has been aimed at a tab. It leaves the active list
 * at send time and lands here, recording WHERE it went: the tab's label and,
 * when the tab has one, the agent session id — the question the history exists
 * to answer, and the one thing that still identifies the conversation once the
 * tab is closed.
 */
export interface SentAgentPrompt {
  id: string;
  message: string;
  created_at: string;
  sent_at: string;
  tab_label: string;
  session_id?: string;
  preface?: string[];
  /** The agent the tab runs (`claude`, `codex`, …) — what the tab *is*, which
   *  still means something once a tab called "Agent 3" is closed. */
  agent?: string;
  /** How the delivery ended. Absent while the prompt is only queued. */
  result?: "delivered" | "missed" | "failed";
  /** The occurrence it was due at, as a local wall-clock key. */
  scheduled_for?: string;
  /** The tags the prompt carried when it was collected. */
  tags?: string[];
  /** Prompt blame: the full hash HEAD pointed at when the prompt was
   *  delivered, and the branch when there was one. Local repos only. */
  commit?: string;
  branch?: string;
  /** The files that changed between the delivery and the agent going idle,
   *  recorded once by `agent_prompt_blame`; `files_at` is when. Absent until
   *  the scheduler has seen the tab idle again. */
  files?: string[];
  files_at?: string;
}

export interface PromptLink {
  id: string;
  from: string;
  to: string;
  kind: "related" | "after";
  target?: string;
}

/** Send-time facts a history entry records. */
export interface SentPromptFacts {
  tabLabel: string;
  sessionId?: string;
  preface?: string[];
  agent?: string;
  result?: SentAgentPrompt["result"];
  scheduledFor?: string;
}

function sentPayload(sent: SentPromptFacts) {
  return {
    tab_label: sent.tabLabel,
    session_id: sent.sessionId ?? null,
    preface: sent.preface ?? [],
    agent: sent.agent ?? null,
    result: sent.result ?? null,
    scheduled_for: sent.scheduledFor ?? null,
  };
}

interface AgentPromptsStore {
  byProject: Record<string, ProjectAgentPrompt[]>;
  historyByProject: Record<string, SentAgentPrompt[]>;
  linksByProject: Record<string, PromptLink[]>;
  loading: Record<string, boolean>;
  load: (projectId: string) => Promise<ProjectAgentPrompt[]>;
  loadHistory: (projectId: string) => Promise<SentAgentPrompt[]>;
  loadLinks: (projectId: string) => Promise<PromptLink[]>;
  /** `tags` undefined leaves an existing prompt's tags alone (the phone edits
   *  text only); an array replaces them, empty included. */
  upsert: (
    projectId: string,
    prompt: { id: string; message: string; tags?: string[] },
  ) => Promise<ProjectAgentPrompt[]>;
  remove: (projectId: string, promptId: string) => Promise<ProjectAgentPrompt[]>;
  reorder: (projectId: string, ids: string[]) => Promise<ProjectAgentPrompt[]>;
  archive: (
    projectId: string,
    promptId: string,
    sent: SentPromptFacts,
  ) => Promise<ProjectAgentPrompt[]>;
  record: (
    projectId: string,
    entry: { id: string; message: string; sent: SentPromptFacts },
  ) => Promise<SentAgentPrompt[]>;
  clearHistory: (projectId: string, entryId?: string) => Promise<SentAgentPrompt[]>;
  link: (projectId: string, link: PromptLink) => Promise<PromptLink[]>;
  unlink: (projectId: string, linkId: string) => Promise<PromptLink[]>;
  /** Record the files a delivered prompt touched (see `agent_prompt_blame`). */
  blame: (projectId: string, entryId: string, since?: string) => Promise<SentAgentPrompt[]>;
  refreshLoaded: () => Promise<void>;
}

export const useAgentPromptsStore = create<AgentPromptsStore>((set, get) => ({
  byProject: {},
  historyByProject: {},
  linksByProject: {},
  loading: {},

  load: async (projectId) => {
    set((state) => ({ loading: { ...state.loading, [projectId]: true } }));
    try {
      const prompts = await invoke<ProjectAgentPrompt[]>("agent_prompts_list", { projectId });
      set((state) => ({ byProject: { ...state.byProject, [projectId]: prompts } }));
      return prompts;
    } finally {
      set((state) => ({ loading: { ...state.loading, [projectId]: false } }));
    }
  },

  loadHistory: async (projectId) => {
    const history = await invoke<SentAgentPrompt[]>("agent_prompt_history_list", { projectId });
    set((state) => ({ historyByProject: { ...state.historyByProject, [projectId]: history } }));
    return history;
  },

  loadLinks: async (projectId) => {
    const links = await invoke<PromptLink[]>("agent_prompt_links_list", { projectId });
    set((state) => ({ linksByProject: { ...state.linksByProject, [projectId]: links } }));
    return links;
  },

  upsert: async (projectId, prompt) => {
    const prompts = await invoke<ProjectAgentPrompt[]>("agent_prompt_upsert", {
      projectId,
      prompt: { id: prompt.id, message: prompt.message, tags: prompt.tags ?? null },
    });
    set((state) => ({ byProject: { ...state.byProject, [projectId]: prompts } }));
    return prompts;
  },

  remove: async (projectId, promptId) => {
    const prompts = await invoke<ProjectAgentPrompt[]>("agent_prompt_delete", { projectId, promptId });
    set((state) => ({ byProject: { ...state.byProject, [projectId]: prompts } }));
    return prompts;
  },

  /**
   * Persist a dragged order. The list is written optimistically before the
   * command answers: the drop already moved the row on screen, and painting it
   * back to where it was for the length of a round trip is what makes a
   * reorder feel like it did not take.
   */
  reorder: async (projectId, ids) => {
    const before = get().byProject[projectId] ?? [];
    const staged = [
      ...ids.map((id) => before.find((prompt) => prompt.id === id)).filter((p): p is ProjectAgentPrompt => !!p),
      ...before.filter((prompt) => !ids.includes(prompt.id)),
    ];
    set((state) => ({ byProject: { ...state.byProject, [projectId]: staged } }));
    try {
      const prompts = await invoke<ProjectAgentPrompt[]>("agent_prompt_reorder", { projectId, ids });
      set((state) => ({ byProject: { ...state.byProject, [projectId]: prompts } }));
      return prompts;
    } catch (cause) {
      set((state) => ({ byProject: { ...state.byProject, [projectId]: before } }));
      throw cause;
    }
  },

  archive: async (projectId, promptId, sent) => {
    const prompts = await invoke<ProjectAgentPrompt[]>("agent_prompt_archive", {
      projectId,
      promptId,
      sent: sentPayload(sent),
    });
    set((state) => ({ byProject: { ...state.byProject, [projectId]: prompts } }));
    await get().loadHistory(projectId).catch(() => []);
    return prompts;
  },

  record: async (projectId, entry) => {
    const history = await invoke<SentAgentPrompt[]>("agent_prompt_record", {
      projectId,
      entry: { id: entry.id, message: entry.message, created_at: null, sent: sentPayload(entry.sent) },
    });
    set((state) => ({ historyByProject: { ...state.historyByProject, [projectId]: history } }));
    return history;
  },

  clearHistory: async (projectId, entryId) => {
    const history = await invoke<SentAgentPrompt[]>("agent_prompt_history_clear", {
      projectId,
      entryId: entryId ?? null,
    });
    set((state) => ({ historyByProject: { ...state.historyByProject, [projectId]: history } }));
    return history;
  },

  link: async (projectId, link) => {
    const links = await invoke<PromptLink[]>("agent_prompt_link_upsert", { projectId, link });
    set((state) => ({ linksByProject: { ...state.linksByProject, [projectId]: links } }));
    return links;
  },

  unlink: async (projectId, linkId) => {
    const links = await invoke<PromptLink[]>("agent_prompt_link_delete", { projectId, linkId });
    set((state) => ({ linksByProject: { ...state.linksByProject, [projectId]: links } }));
    return links;
  },

  blame: async (projectId, entryId, since) => {
    const history = await invoke<SentAgentPrompt[]>("agent_prompt_blame", {
      projectId,
      entryId,
      since: since ?? null,
    });
    set((state) => ({ historyByProject: { ...state.historyByProject, [projectId]: history } }));
    return history;
  },

  refreshLoaded: async () => {
    await Promise.all(Object.keys(get().byProject).map((projectId) => get().load(projectId).catch(() => [])));
    await Promise.all(
      Object.keys(get().historyByProject).map((projectId) => get().loadHistory(projectId).catch(() => [])),
    );
    await Promise.all(
      Object.keys(get().linksByProject).map((projectId) => get().loadLinks(projectId).catch(() => [])),
    );
  },
}));

/**
 * Aim a message at one agent tab now: a one-time schedule at the current
 * minute (see `lib/agentPromptSend`). Finished one-time schedules are pruned
 * first when the tab is at its cap; the count is returned so the caller can say
 * so. The scheduler host hears the change event and delivers at the next idle
 * point.
 */
export async function queuePromptForTab(
  projectId: string,
  scheduleTargetId: string,
  message: string,
  options: { preface?: string[]; now?: Date; id?: string } = {},
): Promise<{ pruned: number }> {
  const now = options.now ?? new Date();
  const schedules = useAgentSchedulesStore.getState();
  const existing = await schedules.load(projectId, scheduleTargetId);
  const prune = schedulesToPruneForSend(existing);
  for (const id of prune) await schedules.remove(projectId, scheduleTargetId, id);
  await schedules.upsert(
    projectId,
    scheduleTargetId,
    buildSendNowSchedule(message, now, options.id ?? crypto.randomUUID(), options.preface),
  );
  return { pruned: prune.length };
}

/**
 * Send a COLLECTED prompt at one tab: queue it, then retire it to the history.
 *
 * One helper rather than two call sites doing it in sequence, because the two
 * halves must not come apart — a prompt delivered but left in the active list is
 * one the user sends twice. The queue is what can fail; the archive is
 * best-effort and never undoes it, since the send has already happened.
 */
export async function sendCollectedPrompt(
  projectId: string,
  target: { scheduleTargetId: string; label: string; sessionId?: string; agent?: string },
  prompt: { id: string; message: string },
  preface?: string[],
): Promise<{ pruned: number }> {
  // The queued schedule carries the PROMPT's id, which is the id its history
  // entry is written under. When the scheduler delivers it, the record it
  // writes lands on that same row and turns "queued" into "delivered" —
  // without the shared id the same prompt would be listed twice.
  const result = await queuePromptForTab(projectId, target.scheduleTargetId, prompt.message, {
    preface,
    id: prompt.id,
  });
  await useAgentPromptsStore
    .getState()
    .archive(projectId, prompt.id, {
      tabLabel: target.label,
      sessionId: target.sessionId,
      preface,
      agent: target.agent,
    })
    .catch(() => []);
  return result;
}

/**
 * Write one scheduled delivery onto the project's history — the record that
 * outlives the rule.
 *
 * It throws rather than swallowing: the caller retires the rule once this
 * succeeds, and a rule deleted after a failed write would take the only
 * account of the delivery with it. A caller that cannot act on the failure
 * ignores it; the scheduler retries on its next tick.
 */
export async function recordScheduledDelivery(
  projectId: string,
  entry: { id: string; message: string; preface?: string[] },
  sent: SentPromptFacts,
): Promise<void> {
  await useAgentPromptsStore.getState().record(projectId, {
    id: entry.id,
    message: entry.message,
    sent: { ...sent, preface: sent.preface ?? entry.preface },
  });
}

File: src-tauri/src/commands/agent_prompts.rs
use tauri::{AppHandle, Emitter};

use crate::{
    schema::agent_prompts::{
        ProjectAgentPrompt, ProjectAgentPromptInput, PromptLink, PromptLinkInput,
        RecordedAgentPromptInput, SentAgentPrompt, SentAgentPromptInput,
    },
    services::agent_prompts,
};

use super::git::run_off_thread;

const CHANGED_EVENT: &str = "agent-prompts-changed";

fn changed(app: &AppHandle) {
    let _ = app.emit(CHANGED_EVENT, ());
}

#[tauri::command]
pub fn agent_prompts_list(project_id: String) -> Result<Vec<ProjectAgentPrompt>, String> {
    agent_prompts::list(&project_id)
}

#[tauri::command]
pub fn agent_prompt_links_list(project_id: String) -> Result<Vec<PromptLink>, String> {
    agent_prompts::links(&project_id)
}

#[tauri::command]
pub fn agent_prompt_link_upsert(
    app: AppHandle,
    project_id: String,
    link: PromptLinkInput,
) -> Result<Vec<PromptLink>, String> {
    let result = agent_prompts::link_upsert(&project_id, link)?;
    changed(&app);
    Ok(result)
}

#[tauri::command]
pub fn agent_prompt_link_delete(
    app: AppHandle,
    project_id: String,
    link_id: String,
) -> Result<Vec<PromptLink>, String> {
    let result = agent_prompts::link_delete(&project_id, &link_id)?;
    changed(&app);
    Ok(result)
}

#[tauri::command]
pub fn agent_prompt_upsert(
    app: AppHandle,
    project_id: String,
    prompt: ProjectAgentPromptInput,
) -> Result<Vec<ProjectAgentPrompt>, String> {
    let result = agent_prompts::upsert(&project_id, prompt)?;
    changed(&app);
    Ok(result)
}

#[tauri::command]
pub fn agent_prompt_delete(
    app: AppHandle,
    project_id: String,
    prompt_id: String,
) -> Result<Vec<ProjectAgentPrompt>, String> {
    let result = agent_prompts::delete(&project_id, &prompt_id)?;
    changed(&app);
    Ok(result)
}

/// Reorder a project's collected prompts — the order a drag left them in. The
/// caller sends the ids in their new order; anything it did not name keeps its
/// relative place after them.
#[tauri::command]
pub fn agent_prompt_reorder(
    app: AppHandle,
    project_id: String,
    ids: Vec<String>,
) -> Result<Vec<ProjectAgentPrompt>, String> {
    let result = agent_prompts::reorder(&project_id, ids)?;
    changed(&app);
    Ok(result)
}

#[tauri::command]
pub fn agent_prompt_history_list(project_id: String) -> Result<Vec<SentAgentPrompt>, String> {
    agent_prompts::history(&project_id)
}

/// Retire a collected prompt to the history after it has been aimed at a tab.
/// Returns the remaining active prompts; the history is reloaded separately.
///
/// Off-thread, like the git commands: the row is stamped with the project's
/// HEAD (`services::prompt_blame`), which spawns git for a local project.
#[tauri::command]
pub async fn agent_prompt_archive(
    app: AppHandle,
    project_id: String,
    prompt_id: String,
    sent: SentAgentPromptInput,
) -> Result<Vec<ProjectAgentPrompt>, String> {
    let result =
        run_off_thread(move || agent_prompts::archive(&project_id, &prompt_id, sent)).await?;
    changed(&app);
    Ok(result)
}

/// Record a delivery the scheduler made. Nothing was collected, so this
/// returns the history rather than the active list; an entry already written
/// under this id (a send-now prompt archived when it was queued) is updated in
/// place with the outcome.
#[tauri::command]
pub async fn agent_prompt_record(
    app: AppHandle,
    project_id: String,
    entry: RecordedAgentPromptInput,
) -> Result<Vec<SentAgentPrompt>, String> {
    let result = run_off_thread(move || agent_prompts::record(&project_id, entry)).await?;
    changed(&app);
    Ok(result)
}

/// Prompt blame: record on one history row the files that changed between
/// its delivery (or `since`, the moment the scheduler submitted the text) and
/// now. The scheduler calls it once, when it sees the tab idle again after a
/// delivery. Returns the project's history.
#[tauri::command]
pub async fn agent_prompt_blame(
    app: AppHandle,
    project_id: String,
    entry_id: String,
    since: Option<String>,
) -> Result<Vec<SentAgentPrompt>, String> {
    let result =
        run_off_thread(move || agent_prompts::blame(&project_id, &entry_id, since.as_deref()))
            .await?;
    changed(&app);
    Ok(result)
}

#[tauri::command]
pub fn agent_prompt_history_clear(
    app: AppHandle,
    project_id: String,
    entry_id: Option<String>,
) -> Result<Vec<SentAgentPrompt>, String> {
    let result = agent_prompts::clear_history(&project_id, entry_id.as_deref())?;
    changed(&app);
    Ok(result)
}
